cirq-core/cirq/optimizers/transformer_primitives.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines primitives for common transformer patterns."""
+
+from collections import defaultdict
+from typing import (
+    cast,
+    Callable,
+    Dict,
+    Hashable,
+    Optional,
+    Sequence,
+    TYPE_CHECKING,
+)
+
+from cirq import circuits, ops, protocols
+from cirq.circuits.circuit import CIRCUIT_TYPE
+
+if TYPE_CHECKING:
+    import cirq
+
+MAPPED_CIRCUIT_OP_TAG = '<mapped_circuit_op>'
+
+
+def _to_target_circuit_type(
+    circuit: circuits.AbstractCircuit, target_circuit: CIRCUIT_TYPE
+) -> CIRCUIT_TYPE:
+    return cast(
+        CIRCUIT_TYPE,
+        circuit.unfreeze() if isinstance(target_circuit, circuits.Circuit) else circuit.freeze(),
+    )
+
+
+def map_moments(
+    circuit: CIRCUIT_TYPE,
+    map_func: Callable[[ops.Moment, int], Sequence[ops.Moment]],
+) -> CIRCUIT_TYPE:
+    """Applies local transformation on moments, by calling `map_func(moment)` for each moment.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Moment, moment_index) to a sequence of moments.
+    """
+    return _to_target_circuit_type(
+        circuits.Circuit(map_func(circuit[i], i) for i in range(len(circuit))), circuit
+    )
+
+
+def map_operations(
+    circuit: CIRCUIT_TYPE,
+    map_func: Callable[[ops.Operation, int], ops.OP_TREE],
+) -> CIRCUIT_TYPE:
+    """Applies local transformations on operations, by calling `map_func(op)` for each op.
+
+    Note that the function assumes `issubset(qubit_set(map_func(op)), op.qubits)` is True.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Operation, moment_index) to a cirq.OP_TREE. If the
```

Comment:
Why is the index `i` needed ?

Reply:
The index `i` is needed for situations when the mapping to be applied on the operation also depends upon which moment the operation is currently present in. For example, when characterizing and compensating for fsim gate angles, the moment index can be useful to determine what compensation to apply.